### ISSUE TITLE
Eng Quality Initiatives

### ISSUE THREAD

**Issue by igorsyl** [stacks-network/stacks-core#3408]
Collecting issues and tasks to improve code quality and developer productivity.

## Code quality issues

- Address `rustc` warnings.
- Move Cargo package from `/testnet/stacks-node` to `/stacks-node`
- Move Cargo workspace member "." to `blockstack-core` directory
- Remove code using `unsafe` keyword
- Investigate issues with libraries `reqwest`, `async-h1` and `tokio`
  - https://discord.com/channels/621759717756370964/730814571517968436/1043188561349971998
- Add the Cargo package `stacks-common` as a workspace member
- Remove the use of `extern crate`
  - https://doc.rust-lang.org/edition-guide/rust-2018/path-changes.html
- Add support for Bitcoin v23

## Test suite reliability

Some tests are slow (run for over one minute) or fail when run with multiple threads.

### Running local tests
```
cargo test -r --workspace -- -Z unstable-options --report-time --format=json --test-threads=1 | tee test.log
```

### Running contiguous integration tests
```
PATH=$PATH:/path-to/bitcoin-22.0/bin
BITCOIND_TEST=1
cargo test -r --workspace -- -Z unstable-options --report-time --format=json --test-threads=1 | tee test-ignored.log 
```

Using the logs files produced above:

### Show failing tests
Run local tests with `--test-threads=16` to check for tests failing due to concurrent runs
```
cat test.log | jq -R 'fromjson?' | jq -scr 'map(select(.type=="test" and .event=="failed")) | .[] | "begintest", .name,.reason,.stdout, "endtest"' 
```

### Show slow tests
```
cat test.log | jq -R 'fromjson?' | jq -sc 'map(select(.type=="test" and .event=="ok")) | sort_by(.exec_time) | .[]'
```

## Break up large code files

```
find . -name "*.rs" -not -path "./target/*" -exec wc -l {} + | sort -n
```

```
7739 ./src/chainstate/stacks/db/transactions.rs
8044 ./src/chainstate/burn/db/sortdb.rs
9934 ./testnet/stacks-node/src/tests/neon_integrations.rs
11136 ./src/chainstate/stacks/db/blocks.rs
11535 ./src/chainstate/stacks/miner.rs
286891 total
```

**Comment by gregorycoppola** [comment]
Intro project idea:
* add comments to the major structs and fields in the codebase
  * explain anything about each field that needs to be known but isn't clear from the name and type
  * explain the role of each struct, and its relationship to the rest of the system

**Comment by cylewitruk** [comment]
@gregorycoppola how the documentation should be written would be a good start.  What's expected, etc.  I'm new to rust, so knowing the things in https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html would be a good precursor.

**Comment by jcnelson** [comment]
Filling in some context here.

> Address rustc warnings.

Such as?  I think we got this down to, like, 1.

> Some tests are slow (run for over one minute) or fail when run with multiple threads.

Yes -- some tests simulate running multiple instances of the peer network in order to test some emergent behaviors (e.g. "the neighbor discovery algorithm eventually discovers all other neighbors").  This not only requires binding on ports (which complicates running many of these tests in parallel), but also can take time.

Integration tests are time-consuming in part because they spin up a full Bitcoin node and sometimes multiple Stacks runloops, in order to simulate multiple nodes communicating with one another.

> Remove code using unsafe keyword

We're in a great place with this in our code, but not in the dependencies.  But the use of `unsafe` in dependencies puts the whole node at risk, as we discovered today with `reqwest`.

> Investigate issues with libraries reqwest, async-h1 and tokio

I'd prefer just removing them outright if at all possible.  Why bring in dependencies that segfault?

> Break up large code files

These files are large because they have lots of tests in them.  We could separate the tests out into their own files.  This was already done in `next` for `miner.rs`, for example.

**Comment by jcnelson** [comment]
One thing we could add is to start turning on Clippy warnings, and getting those fixed.

**Comment by cylewitruk** [comment]
Including a [.editorconfig](https://editorconfig.org/) could help to standardize..

**Comment by cylewitruk** [comment]
> Such as? I think we got this down to, like, 1.

```text
warning: unused variable: `stream`
   --> stacks-common\./src\util\log.rs:351:11
    |
351 | fn isatty(stream: Stream) -> bool {
    |           ^^^^^^ help: if this is intentional, prefix it with an underscore: `_stream`
    |
    = note: `#[warn(unused_variables)]` on by default
```
Can't this just be renamed to `_stream` so I stop seeing yellow? 😅 

> Yes -- some tests simulate running multiple instances of the peer network in order to test some emergent behaviors (e.g. "the neighbor discovery algorithm eventually discovers all other neighbors"). This not only requires binding on ports (which complicates running many of these tests in parallel), but also can take time.
>
>Integration tests are time-consuming in part because they spin up a full Bitcoin node and sometimes multiple Stacks runloops, in order to simulate multiple nodes communicating with one another.

Possibly differentiate between unit & integration tests so that they can be run separately? (i'm just assuming that cargo supports multiple build targets, still learning this area..)

**Comment by jcnelson** [comment]
> Can't this just be renamed to _stream so I stop seeing yellow? ￼

What settings are you using?  I see no warnings at all when I run `cargo build`, and I don't have anything non-default set.

> Possibly differentiate between unit & integration tests so that they can be run separately? (i'm just assuming that cargo supports multiple build targets, still learning this area..)

This is already the case.  In fact, each integration test runs in parallel.

**Comment by cylewitruk** [comment]
@jcnelson 
>What settings are you using? I see no warnings at all when I run cargo build, and I don't have anything non-default set.

Just a plain `cargo build` on a fresh `next` branch.  Has been around for me since I started with bitwise, just assumed it was for a reason.  Not using anything non-default that I know of.  Maybe just another Windows ghost..
